### DISCUSSION TITLE
How to link user from budibase->users to postgres table?

### DISCUSSION THREAD

**Discussion by upgreydd** [Budibase/budibase#5847] [Help] [answered]
Hello. I've created a table (in postgres) where I need `created_by` field. so I've put `{{ Current User._id }}` into a `Save Row` action and I see a unique ID there for the specific user, but I can't link it in any known by me way. How I can achieve that? I think the relation between any kind of database should allow access `Users` table in budibase.

**Comment by melohagan** [comment]
Hey @Upgreydd 

Are you trying to populate the auto-generated create by field in your table?
![Screenshot 2022-05-12 at 13 57 45](https://user-images.githubusercontent.com/101575380/168079945-c4498bee-736d-4d22-8778-c8a8bfc3f59d.png)
 
This can't be set manually, and should be automatically set whenever a new row is created in the UI by a user.

**Reply by melohagan** [reply]
Sorry I just noticed you're using PostgreSQL. I don't think there is a way to create a hard relationship between the internal Users table and an external datasource. 

What is your use case?

**Reply by upgreydd** [reply]
I'm just creating a record in the table and need to save who did the action on the table.
I know I can declare a new screen and call the user data provider and filter it - but it's not desired use case - would be better to bind it in the table. I think such linking should be possible via relation column.

**Reply by melohagan** [reply]
I understand, but I don't think this can be feasibly done. What you're asking for is to create a relationship between a NoSQL database table and an RDMS database table. 

Maybe you could try using the custom query instead of a data provider and filter. That way you could pass the `{{ Current User._id }}` through to a SELECT WHERE query. 
https://docs.budibase.com/docs/custom-queries

**Reply by upgreydd** [reply]
@melohagan custom queries are impossible in budibase db. Probably instead of linking data to users table I should write the whole string record in the table from user who did the action - I see no other way unfortunately :/

**Reply by melohagan** [reply]
Yes I think that is the only way at the moment. There is no way to do custom user management with your own DB. Perhaps that's something that could be brought up when the user management is being revamped.
It's on the roadmap for Q3: https://github.com/Budibase/budibase/projects/10

**Comment by Cityjohn** [comment]
@melohagan 

Does this mean there is no way to build for example, a user option dropdown list in forms which are based in an SQL datasource?

Since you can automate insert/update/delete it should be possible to essentially build a complete cascading system, and even using just insert automation would work since ID's are incremental and unique.

Wouldn't it be a very easy workaround to simply create an automation that copies the entire user table into the SQL database so that the relationship can be offloaded?

**Reply by Cityjohn** [reply]
> @melohagan
> 
> Does this mean there is no way to build for example, a user option dropdown list in forms which are based in an SQL datasource?
> 
> Since you can automate insert/update/delete it should be possible to essentially build a complete cascading system, and even using just insert automation would work since ID's are incremental and unique.
> 
> Wouldn't it be a very easy workaround to simply create an automation that copies the entire user table into the SQL database so that the relationship can be offloaded?

Oh noes,

Budibase automation doesn't include changes to the user table :(( @melohagan can this be remedied?

**Reply by melohagan** [reply]
Hey @Cityjohn I think you could use OpenID Connect for supporting your own IDP and have your own users database: https://docs.budibase.com/docs/openid-connect

I think at the moment that would also create entries in the Budibase users table, and the roles are still based on that table.

**Comment by apparentlyjeff** [comment, marked as answer]
It's not the most ideal solution,  but my workaround was to create my own user table in my database (MySQL, but same process) that somewhat duplicates the Budibase user table.  I set the email/username field as unique, since Budibase doesn't seem to handle checking for non existence very well. I then created a cron automation in budibase that simply loops through the Users table and inserts into my DB table every hour.  The unique property of that email field prevents duplicate records , so I didn't have to worry about doing any lookup. 

My whole reasoning for doing this was mainly to be able to join user information in DB queries. My application basically uses my own user table for everything it displays and just uses the Budibase Users table for application level permissions. 

If there's need for a change to the username/email, it will need to be a manual update to the tables to keep them in sync.  However that situation is less frequent than the need to join user data, so I'm willing to work with it.

**Reply by melohagan** [reply]
Thanks for sharing!

**Reply by vishal-bluebash** [reply]
@apparentlyjeff do you have any code references on how you have done it, I am trying to do same

**Reply by apparentlyjeff** [reply]
You'll want to create a user table in your own database.  In that table, you'll want a field for the user_id, and set it up with a unique index, so any duplicate insertions will fail.  You can, obviously, also add your own additional fields if you desire.

Then set up an automation in Budibase similar to below.  It simply queries the budibase users table and inserts each record into your own databaes.  There's no real logic to it:
![image](https://github.com/Budibase/budibase/assets/15859008/9568c81d-0ab4-46ab-828d-de4630fa1c46)

**Reply by vishal-bluebash** [reply]
Thank you @apparentlyjeff I don't see image attached

**Reply by apparentlyjeff** [reply]
Yeah, sorry, I got interrupted while posting the image and uploaded the wrong one.

**Reply by vishal-bluebash** [reply]
Hi @apparentlyjeff Thank you again., looks like automation is working, but I don't see indexes section my external source is Postgresql

<img width="1512" alt="Screenshot 2023-06-11 at 3 48 28 AM" src="https://github.com/Budibase/budibase/assets/43650278/96478e22-542b-4ab9-9f9a-305ea341ba77">

Also, is there any easy way to set up automatic created_at and updated_at fields (I know how to do it directly in database but not sure how can we do it in budibase postgresl)

**Reply by apparentlyjeff** [reply]
You'll want to do both the index and the default values in Postgres directly, not through the Budibase interface.  Then return to budibase, click on the "Postgres" data source and finally click the "Fetch Tables" button.  Budibase will then pull in the external details and you'll be good.

**Reply by vishal-bluebash** [reply]
@apparentlyjeff okay make sense, Thank you

**Reply by vishal-bluebash** [reply]
<img width="1205" alt="Screenshot 2023-06-24 at 12 32 04 AM" src="https://github.com/Budibase/budibase/assets/43650278/16f485f3-f85d-40fb-b6c1-a340aaa58ac5">

<img width="1004" alt="Screenshot 2023-06-24 at 12 31 52 AM" src="https://github.com/Budibase/budibase/assets/43650278/a4f1ebe5-18c9-4a76-875d-b15bd6a8f94c">

I have copied users into postgres, now I have challenge I want to scope results based on assignees and assignee is reference from my postgres table. in current user I am getting user from budibase table.

I have made query to return current user id from postgres w.r.t current user logged in through budibase.  But when I apply filter and do dynamic binding it is not working . 

@apparentlyjeff  any help will be appreciated

**Reply by vishal-bluebash** [reply]
I solved it by the way. A copy of the user table that we have in Postgres, we created an email as the primary key.
